### PR TITLE
CRNS-261: Improving experience around overriding gesture/touchable handlers

### PR DESCRIPTION
**Pull request by vishalnarkhede**
**NOT YET READY**

## TODO

- [x] test android
- [x] test expo
- [x] test onPress and onPressIn handlers on attachments
- [x] check abc
- [x] test threads
- [x] test messageActions

## Objective

- add support for `animatedLongPress` prop on Channel component
- bring consistency on gesture/touchable handlers such as onLongPress, onPress, onPress etc, that are overridable through props - either on Channel component, Message component or individual attachment sub components.

## Description

 With current implementation, we don't have consistency in terms of what parameters do these functions have access to, and also you don't have access to message actions within these functions. Especially 2nd requirement is quite common, e.g. I want to open message overlay on double top (whatsapp), I want to toggle reaction on double tap (instagram), I want to open my custom actionsheet (slack). We have handled "double-tap-reaction" explicitly already, but lets make it more scalable.

  With proposed implementation, user will have access to following props on `Channel` component
  - `onLongPressMessage`
  - `onPressInMessage`
  - `onPressMessage`
  - `onDoubleTapMessage`
  
  (And this leaves us room to introduce new gestures in future e.g,  swipe to edit)

  All these props will accept following param:

  ```jsx
  <Channel
    onLongPressMessage={(
      // Keeping this object, so that we have option to introduce more params here, if needed in future
      {
        message,
        actionHandlers: {
          editMessage,
          deleteMessage,
          toggleReaction
          ...
        },
        defaultGestureHandler, // default action for corresponding gesture handlger
        event // original event that triggered this action
    }) => {
      // Whatever you would like to do here
    }}
    onPressMessage={/** same */}
    onPressInMessage={/** same */}
  />
	   
  ```

  Message component will accept following props:

  ```jsx
   <Channel
    Message={props => (
      <Message
        {...props}
        onLongPress={(event, defaultGestureHandler) => { /** your custom action */ }}
        onPressIn={(event, defaultGestureHandler) => { /** your custom action */ }}
        onPress={(event, defaultGestureHandler) => { /** your custom action */ }}
    )}
  />

  ```

#### Double tap reaction (instagram, fb messenger)

```jsx
<Channel
  onDoubleTapMessage={({ actionHandlers }) => actionHandlers.toggleReaction('love')}
/>
```

#### Double tap to open message overlay (whatsapp, iMessage)

```jsx
<Channel
  onDoubleTapMessage={({ actionHandlers }) => actionHandlers.showMessageOverlay()}
/>
```

#### long press to open some custom bottomsheet (slack)

```jsx
<Channel
  onLongPressMessage={({ message, actionHandlers }) => { /** go crazy here */ }}
/>
```

#### track analytics

```jsx
<Channel
  onLongPressMessage={({ defaultGestureHandler }) => {
    trackingCall();
    defaultGestureHandler();
  }}
/>
```

And infinite other possibilities!!

### CODE REVIEW

File: src/components/Message/Message.tsx [hunk 1]
```diff
@@ -417,14 +470,22 @@ const MessageWithContext = <
...
-      pressActive.value = false;
-      cancelAnimation(scale);
-      scale.value = withTiming(1, { duration: 100 });
-      goToMessage(quotedMessage.id);
+    } else if (quotedMessage) {
+      onPressQuotedMessage(quotedMessage);
     }
   };
 
+  const onPressQuotedMessage = (
```
Comment (nhannah) [inline src/components/Message/Message.tsx:478]: Should move up so it's created before called just for logical reading.

File: src/components/Message/Message.tsx [hunk 2]
```diff
@@ -575,6 +636,83 @@ const MessageWithContext = <
...
+          );
+        }
+      }
+    } catch (err) {
+      console.log(err);
+    }
+  };
+
+  const handleResendMessage = () => {
+    const messageWithourReservedFields = removeReservedFields(message);
```
Comment (nhannah) [inline src/components/Message/Message.tsx:666]: spelling

File: src/components/Message/Message.tsx [hunk 3]
```diff
@@ -575,6 +636,83 @@ const MessageWithContext = <
...
+    return retrySendMessage(
+      messageWithourReservedFields as MessageResponse<At, Ch, Co, Me, Re, Us>,
+    );
+  };
+
+  const handleReplyMessage = () => {
+    setQuotedMessageState(message);
+  };
+
+  const isMuted = (client.mutedUsers || []).some(
```
Comment (nhannah) [inline src/components/Message/Message.tsx:677]: Can we remove the repeat of this const later using this one?

File: src/components/Message/Message.tsx [hunk 4]
```diff
@@ -809,27 +920,12 @@ const MessageWithContext = <
...
-              'id',
-              'last_message_at',
-              'member_count',
-              'type',
-              'updated_at',
-            ];
-            reserved.forEach((key) => {
-              delete retryMessage[key];
-            });
+            const messageWithourReservedFields = removeReservedFields(message);
```
Comment (nhannah) [inline src/components/Message/Message.tsx:923]: spelling

File: src/components/Message/Message.tsx [hunk 5]
```diff
@@ -124,6 +125,45 @@ const prefetchImage = ({
...
+    | 'message'
+    | 'messageContent'
+    | 'messageReplies'
+    | 'reactionList'
+    | 'textLink'
+    | 'textMention';
+  event?: GestureResponderEvent;
+};
+
+export type MessageGestureHandlerPayload<
```
Comment (nhannah) [inline src/components/Message/Message.tsx:144]: I think the naming here might be a bit confusing as it doesn't relate to the gesture handlers

File: src/components/Message/Message.tsx [hunk 6]
```diff
@@ -958,52 +1054,44 @@ const MessageWithContext = <
...
+    editMessage: handleEditMessage,
+    reply: handleReplyMessage,
+    resendMessage: handleResendMessage,
+    showMessageOverlay,
+    toggleBanUser: handleToggleBanUser,
+    toggleMuteUser: handleToggleMuteUser,
+    toggleReaction: handleToggleReaction,
+  };
+
   const onLongPressMessage =
```
Comment (nhannah) [inline src/components/Message/Message.tsx:1068]: runInJS functions aren't providing these params

**Review by nhannah — CHANGES_REQUESTED** [review]
Ok so there. are some changes needed but I also need to look more at the Message component as a whole, lot of changes there so I want to make sure I follow it all correctly.

File: src/components/Attachment/Attachment.tsx [hunk 7]
```diff
@@ -245,7 +229,7 @@ export const Attachment = <
     PropFileAttachment || ContextFileAttachment || FileAttachmentDefault;
   const Gallery = PropGallery || ContextGallery || GalleryDefault;
   const Giphy = PropGiphy || ContextGiphy || GiphyDefault;
-  const onPressIn = propOnPressIn || onPressInMessage;
+  const onPressIn = propOnPressIn;
```
[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/Attachment/Attachment.tsx:232]: "don't" we "not" ... you confused me with double negation 😂

But you are right, we don't need it :)

File: src/components/Channel/Channel.tsx [hunk 8]
```diff
@@ -215,6 +215,7 @@ export type ChannelPropsWithContext<
     Pick<
       MessagesContextValue<At, Ch, Co, Ev, Me, Re, Us>,
       | 'additionalTouchableProps'
+      | 'animatedLongPress'
```
[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/Channel/Channel.tsx:218]: This is to allow `animatedLongPress` on Channel component. Message component access it from here!!

File: src/components/Message/Message.tsx (hunk 6 above)
Comment (vishalnarkhede) [reply src/components/Message/Message.tsx:1068]: defaults handle it:

```js
            defaultHandler: payload?.defaultHandler || showMessageOverlay,
            emitter: payload?.emitter || 'message',
            event: payload?.event,

```

File: src/components/Message/Message.tsx (hunk 5 above)
Comment (vishalnarkhede) [reply src/components/Message/Message.tsx]: How about `MessageTouchableHandlerPayload`?

File: src/components/Message/Message.tsx (hunk 4 above)
Comment (vishalnarkhede) [reply src/components/Message/Message.tsx]: Done!!

File: src/components/Message/Message.tsx (hunk 3 above)
Comment (vishalnarkhede) [reply src/components/Message/Message.tsx:677]: Done!!

File: src/components/Message/Message.tsx (hunk 2 above)
Comment (vishalnarkhede) [reply src/components/Message/Message.tsx]: Done!!

File: src/components/Message/Message.tsx (hunk 1 above)
Comment (vishalnarkhede) [reply src/components/Message/Message.tsx]: Done!!

File: src/components/Message/MessageSimple/MessageContent.tsx [hunk 9]
```diff
@@ -364,6 +381,7 @@ export const MessageContentWithContext = <
               <Reply styles={{ messageContainer: { maxWidth: vw(60) } }} />
             </View>
           )}
+          {/* @ts-ignore */}
```
[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/Message/MessageSimple/MessageContent.tsx]: I was using some dev release of typescript apparently on vs code!!

[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/Message/MessageSimple/MessageContent.tsx]: Removed it now!!

File: src/components/Message/MessageSimple/MessageReplies.tsx [hunk 10]
```diff
@@ -39,7 +39,9 @@ const styles = StyleSheet.create({
     marginTop: 8,
   },
   curveContainer: {
+    alignItems: 'center',
```
[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/Message/MessageSimple/MessageReplies.tsx]: Pretty sure I saw some wierd alignment issue there. but I can't see it anymore after syncing with master after long time. Removed it!!

**Comment by vishalnarkhede** [comment]
> Ok so there. are some changes needed but I also need to look more at the Message component as a whole, lot of changes there so I want to make sure I follow it all correctly.

Thanks @nhannah !! Have made the changes and added comments to some questions

File: src/components/Attachment/Attachment.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (nhannah) [reply src/components/Attachment/Attachment.tsx]: hahah my bad

File: src/components/Channel/Channel.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (nhannah) [reply src/components/Channel/Channel.tsx:219]: Yeah it's also on the other context though so we need to remove it from there then as it's never accessed from there.

File: src/components/Message/Message.tsx (hunk 5 above)
Comment (nhannah) [reply src/components/Message/Message.tsx]: Sounds good, gesture just seemed off.

File: src/components/Message/Message.tsx (hunk 6 above)
Comment (nhannah) [reply src/components/Message/Message.tsx:1064]: ah ok event just undefined then cool

File: src/components/Channel/Channel.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (vishalnarkhede) [reply src/components/Channel/Channel.tsx:219]: Makes sense!! Removed it now

File: src/components/Message/MessageSimple/MessageReplies.tsx [hunk 11]
```diff
@@ -41,6 +41,7 @@ const styles = StyleSheet.create({
   },
   curveContainer: {
     flexDirection: 'row',
+    justifyContent: 'center',
```
Comment (nhannah) [inline src/components/Message/MessageSimple/MessageReplies.tsx:44]: Don't need this either from what I see

Comment (vishalnarkhede) [reply src/components/Message/MessageSimple/MessageReplies.tsx]: Done!!

**Review by nhannah — APPROVED** [review]
Ok if you tested out everything is still working right and the overrides all work I think we are good to go, we need to update the double tap example in the wiki to reflect the changes (and any other spots we have those noted).

**Comment by vishalnarkhede** [comment]
> Ok if you tested out everything is still working right and the overrides all work I think we are good to go, we need to update the double tap example in the wiki to reflect the changes (and any other spots we have those noted).

Thats great. Tomorrow morning, I will give it one more round of testing, just to make sure nothing was missed after syncing master. If all looks good, will merge it :)

**Pull request merged**